File: src/Products/PinoyBusinessPOS/ExItS.PinoyBusinessPOS.React/src/features/inventory/transfer-exception-custody-policy.ts
import type { InventoryTransferDamagedCustodyDecisionCode } from "@/api/pos/pos-inventory-transfer-client";

export const OTHER_REASONS_FORCE_RETURN = ["WrongItem", "WrongVariant", "Expired"] as const;

export type OtherReasonForceReturnCode = (typeof OTHER_REASONS_FORCE_RETURN)[number];

export function requiresActualProduct(reasonCode: string): boolean {
  return reasonCode === "WrongItem" || reasonCode === "WrongVariant";
}

/** Wrong item/variant source receive restores sellable stock; no inspection UI. */
export function restoresDirectlyToSellableOnSourceReceive(reasonCode: string): boolean {
  return requiresActualProduct(reasonCode);
}

/** Wrong item/variant cannot claim the expected transfer product as the actual SKU. */
export function isActualProductSameAsExpected(
  reasonCode: string,
  expectedProductId: string | null | undefined,
  actualProductId: string | null | undefined,
): boolean {
  if (!requiresActualProduct(reasonCode)) {
    return false;
  }
  const expected = expectedProductId?.trim();
  const actual = actualProductId?.trim();
  if (!expected || !actual) {
    return false;
  }
  return expected.toLowerCase() === actual.toLowerCase();
}

export function forcesReturnToSource(reasonCode: string): boolean {
  return (OTHER_REASONS_FORCE_RETURN as readonly string[]).includes(reasonCode);
}

export function resolveDefaultOtherCustody(
  reasonCode: string,
): InventoryTransferDamagedCustodyDecisionCode {
  if (forcesReturnToSource(reasonCode)) {
    return "ReturnToSource";
  }
  return "KeepAtDestination";
}

File: src/Products/PinoyBusinessPOS/ExItS.PinoyBusinessPOS.Domain/Inventory/TransferExceptionCustodyPolicy.cs
using ExItS.PinoyBusinessPOS.Domain.Common;
using ExItS.PinoyBusinessPOS.Domain.Purchasing;

namespace ExItS.PinoyBusinessPOS.Domain.Inventory;

/// <summary>
/// Authoritative custody rules for transfer "Other" discrepancy reasons.
/// Follow-up (replacement vs accept shortage) remains independent of custody.
/// </summary>
public static class TransferExceptionCustodyPolicy
{
    public static bool RequiresActualProduct(string reasonCode) =>
        reasonCode is ReceiveDiscrepancyOtherReason.WrongItem
            or ReceiveDiscrepancyOtherReason.WrongVariant;

    /// <summary>
    /// Wrong item/variant returns restore directly to source sellable stock (no inspection).
    /// </summary>
    public static bool RestoresDirectlyToSellableOnSourceReceive(string reasonCode) =>
        RequiresActualProduct(reasonCode);

    public static bool ForcesReturnToSource(string reasonCode) =>
        reasonCode is ReceiveDiscrepancyOtherReason.WrongItem
            or ReceiveDiscrepancyOtherReason.WrongVariant
            or ReceiveDiscrepancyOtherReason.Expired;

    public static bool AllowsKeepAtDestination(string reasonCode) =>
        reasonCode is ReceiveDiscrepancyOtherReason.PackagingIssue
            or ReceiveDiscrepancyOtherReason.QualityIssue
            or ReceiveDiscrepancyOtherReason.Other;

    public static bool RequiresDescription(string reasonCode) =>
        string.Equals(reasonCode, ReceiveDiscrepancyOtherReason.Other, StringComparison.Ordinal);

    /// <summary>
    /// Resolves custody for an other-qty wave. Throws when Keep is requested for locked reasons
    /// or when a choosable reason omits a decision.
    /// </summary>
    public static InventoryTransferExceptionCustodyDecision ResolveDecision(
        string reasonCode,
        InventoryTransferExceptionCustodyDecision? requested)
    {
        if (ForcesReturnToSource(reasonCode))
        {
            if (requested is InventoryTransferExceptionCustodyDecision.KeepAtDestination)
            {
                throw new DomainException(
                    DomainErrorCodes.InvalidInventoryTransferExceptionCustodyDecision,
                    $"{reasonCode} requires return to source; keep at destination is not allowed.");
            }

            return InventoryTransferExceptionCustodyDecision.ReturnToSource;
        }

        if (!AllowsKeepAtDestination(reasonCode))
        {
            throw new DomainException(
                DomainErrorCodes.InvalidInventoryTransferOtherReason,
                $"Unknown other discrepancy reason '{reasonCode}'.");
        }

        if (requested is null)
        {
            throw new DomainException(
                DomainErrorCodes.InvalidInventoryTransferExceptionCustodyDecision,
                "Exception custody decision is required for this other reason.");
        }

        return requested.Value;
    }
}
